File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # Swap along the row from top and bottom moving towards each other
        top, bottom = 0, len(matrix) - 1
        while top < bottom:
            matrix[top], matrix[bottom] = matrix[bottom], matrix[top]
            top += 1
            bottom -= 1

        # Fix diagonal and swap elements on opposite sides with each other
        idx = 1
        while idx < len(matrix):
            i, j = idx, 0
            while i < len(matrix):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
                i += 1
                j += 1
            idx += 1

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        int l = 0, r = n - 1;
        while (l < r) {
            for (int i = 0; i < n; i++) {
                int tmp = matrix[l][i];
                matrix[l][i] = matrix[r][i];
                matrix[r][i] = tmp;
            }
            l++;
            r--;
        }

        int diagonal = 0;
        while (diagonal < n ) {
            for (int i = diagonal; i < n; i++) {
                int tmp = matrix[diagonal][i];
                matrix[diagonal][i] = matrix[i][diagonal];
                matrix[i][diagonal] = tmp;
            }
            diagonal++;
        }
    }
    
}